api/types/worker.proto

```diff
@@ -19,9 +19,9 @@ message GCPolicy {
 	int64 keepDuration = 2;
 	repeated string filters = 4;
 
-	int64 minStorage = 5;
-	// maxStorage was renamed from freeBytes
-	int64 maxStorage = 3;
+	// minStorage was renamed from freeBytes
```

Comment:
@jedevc I switched these around. Minimum makes more sense afaics because the minimum value is independent from the new "free" filter, while "max" works in combination of the new "free" field.

Reply:
I'm really confused by this - it's the other way round? See the doc comment here:

https://github.com/moby/buildkit/blob/00c5f6e09c2fa889604b21197dd6601fd9739e2b/cmd/buildkitd/config/config.go#L169-L179

`MaxStorage` is exactly what will be used by `FreeBytes`, I essentially just translated `FreeBytes` to `MaxStorage`. It's the maximum amount of storage that the policy is allowed to consume. `MinStorage` is the number that works with the new `Free` field, the total will never be brought below that line. See the definition of [`calculateKeepBytes`](https://github.com/moby/buildkit/blob/00c5f6e09c2fa889604b21197dd6601fd9739e2b/cache/manager.go#L1081-L1103) - we start with the max (the old `keepBytes`, reduce it if we need to free, but never go below min storage).

Essentially (I made a diagram when working on this originally):

![Scanned_20240927-1217](https://github.com/user-attachments/assets/a82bc0d4-0e35-4048-9c33-6f88c95ea420)

I don't think we should change this, and surely if we do, we need a change in logic *everywhere* (including the actual gc logic in `cache/manager.go`)? Essentially, this change will take an old client's `freeBytes` value, and convert it into `minStorage`, which will then mean that the `maxStorage` is uncapped, and buildkit will fill the disk.

Reply:
If you set `min` to `x` and leave `free` and `max` to 0, or if you set `max` to `x` and leave `free` and `min` to 0, they will behave the same way. Difference only appears if some of the new variables are set as well.


>  I essentially just translated FreeBytes to MaxStorage.

You mean "KeepBytes" to "MaxStorage" I think.

> MinStorage is the number that works with the new Free field, the total will never be brought below that line. 

If you set minstorage on its own then it behaves like `keepBytes` before (as would max on its own). If you set minstorage together with `free` then it still behaves like a minimum value, meaning value of free can't take the leftover cache below that value. If you set `max` together with `free` then the value of `free` can redefine that maximum to a lower value, essentially meaning the `max` defines a maximum value, but only if `free` config doesn't redefine it to a lower value instead. That to me is the new behavior.

> which will then mean that the maxStorage is uncapped, and buildkit will fill the disk.

If you set `min` to a value, and `max` to 0, that does not mean that storage is uncapped. It means that `min` is the storage value. This is in your implementation, not changed by this PR.

Looking over the code again, there is a bug in https://github.com/moby/buildkit/blob/185751bdc010f1f8bf1192d6e9b4ec19552d5741/cache/manager.go#L1048 though. `totalSize` needs to be computed if any of the new storage properties is set, not just `MaxStorage` (or at least set any time `calculateKeepBytes()` returns a non-zero value.

Reply:
> `totalSize` needs to be computed if any of the new storage properties is set, not just `MaxStorage` (or at least set any time `calculateKeepBytes()` returns a non-zero value.

makes sense to me.

> If you set `min` to a value, and `max` to 0, that does not mean that storage is uncapped. It means that `min` is the storage value. This is in your implementation, not changed by this PR.

So assigning the old value of `KeepBytes` to *either* `MinStorage` or `MaxStorage` works then, I misunderstood. Both cases end up as computing the value of `calculateKeepBytes` as the same value.

That means we can pick either to assign the old value to either of them, it's just a matter of preference.

I'd argue assigning to `MaxStorage` makes more sense - if you set `keepBytes` and `free` (an edge case tbf, since users using `free` should use `min`/`max`), then I'd *expect* the result to be the same as before the `free` was introduced, but now, the additional constraint is that we will always leave `free` bytes free. Trying to explain that the value of `free` set here isn't actually accurate, and might not always be followed because `keepBytes` is the minimum used feels trickier to explain.

I think we should pick the option that feels easier to explain, since from a behavior point of view they're similar. In your original explanation https://github.com/moby/buildkit/pull/5079#issuecomment-2192315330, the point that lines up most with the deprecated `keepBytes` is `maxStorage` IMHO.

Reply:
Potentially we could avoid this entirely if we forbid setting `KeepBytes` with `Free`?

Then it doesn't matter how this is translated, and we can translate it at the protobuf layer to `Min` or to `Max`, it doesn't matter?

I think there's potentially too much ambiguity in what `keep` and `free` should mean, so we should force users to be explicit and use the new `Min` and `Max` instead.

Reply:
> I think we should pick the option that feels easier to explain,

I think "minimum allowed storage" is quite a weird concept to understand. That's why I started to introduce lingo of "reserved storage" https://github.com/moby/buildkit/pull/5359/files#diff-df23634beef4b740f3608310b8fcacffe6db412833983980b1efadf34ed3873fR95 https://github.com/moby/buildkit/pull/5359/files#diff-23d73016cb11ae482c811c818fcbd7f701900ee59bb4a6fe541a7620180a4e5bR155 . 

Reserved means storage that is reserved for BuildKit to use and can not be taken away by other applications or config fields. On its own, it behaves the same as the old "keepstorage" field. In addition to "reserved storage", in the new version, the user can also define "free" and "max" that are new fields that define how BuildKit can use additional space on top of the "reserved storage" in case that storage is unused anyway by other applications.

@dvdksn wdyt? We should also decide what flag names we want to keep in `docker buildx prune`.

Reply:
I'm happy with those changes - would have been good to flesh this out in #5079 though.

> I think "minimum allowed storage" is quite a weird concept to understand. That's why I started to introduce lingo of "reserved storage" 

I think this is good, but we should be consistent and rename it everywhere, not just in the user facing code.

If we have "reserved storage", having "maximum storage"  does feel a little odd (no contrasting "minimum storage"). Maybe "reserved storage" and "total storage"? Or "reserved storage" and "available storage"?

Docs-wise, I still think this is tricky to understand, I think having a diagram that draws out where all the numbers are in relation to each other and what they mean would be useful (but I've struggled to come up with one myself).

Reply:
> If we have "reserved storage", having "maximum storage" does feel a little odd (no contrasting "minimum storage"). Maybe "reserved storage" and "total storage"? Or "reserved storage" and "available storage"?

Total seems more like what buildkit uses, not config. I think `max` could be "Storage limit". "Free" is not ideal as well to understand the meaning, but don't have good ideas atm.

Reply:
My suggestion for `buildx prune` is

```
      --reserved-storage bytes    Amount of disk space always allowed to keep for cache
      --free-storage-limit bytes   Target amount of free disk space after pruning
      --max-storage-limit bytes   Maximum amount of disk space allowed to keep for cache
```

Another interesting could be

```
--prune-storage bytes Target amount of disk space to be reclaimed during prune
```

In variables that would mean:

```
type GCConfig struct {
	GC *bool `toml:"gc"`
	// Deprecated: use GCReserved instead
	GCKeepStorage DiskSpace  `toml:"gckeepstorage"`
	GCReserved    DiskSpace  `toml:"gcreserved"`
	GCMax         DiskSpace  `toml:"gcmax"`
	GCFree        DiskSpace  `toml:"gcfree"`
	GCPolicy      []GCPolicy `toml:"gcpolicy"`
}
```

```
type GCPolicy struct {
	All     bool     `toml:"all"`
	Filters []string `toml:"filters"`

	KeepDuration Duration `toml:"keepDuration"`

	// KeepBytes is the maximum amount of storage this policy is ever allowed
	// to consume. Any storage above this mark can be cleared during a gc
	// sweep.
	//
	// Deprecated: use ReservedSpace instead
	KeepBytes DiskSpace `toml:"keepBytes"`

	// ReservedSpace is the minimum amount of disk space this policy is guaranteed to retain.
	// Any usage below this threshold will not be reclaimed during garbage collection.
	ReservedSpace DiskSpace `toml:"reservedSpace"`

	// MaxSpace is the maximum amount of disk space this policy is allowed to use.
	// Any usage exceeding this limit will be cleaned up during a garbage collection sweep.
	MaxSpace DiskSpace `toml:"maxSpace"`

	// FreeSpace is the target amount of free disk space the garbage collector will attempt to leave.
	// However, it will never let the available space fall below MinSpace.
	FreeSpace DiskSpace `toml:"freeSpace"`
}
```

Please confirm.

Reply:
I really like `reservedSpace` 👍🏻 

I'm way out of the loop on this but it did take me a while to understand the relation between `max` and `free`. I wonder if naming these options differently would help. For example, emphasizing that `free` refers to a minimum target of space to leave free. Maybe:

- `reservedSpace`
- `maxUsedSpace`
- `minFreeSpace`

For `buildx prune`, maybe it would be good if the options aligned with the buildkit terminology?

```
--reserved-space bytes      Amount of disk space always allowed to keep for cache
--min-free-space bytes      Target amount of free disk space after pruning
--max-used-space bytes      Maximum amount of disk space allowed to keep for cache
```

I don't exactly understand `--prune-storage`. Is it different from `--min-free-space`?

Reply:
> I don't exactly understand --prune-storage. Is it different from --min-free-space?

The value here is not the target disk space after pruning, but how much data to prune. Eg. lets say I fill up my disk and then think "I want to delete 5GB of build cache". There is no equivalent for this field in GC policy, it would only be a filter for the prune command (and it is currently not implemented).

Reply:
OK, I should have understood that 😅 Would it only take bytes or do you imagine units and/or percent too? Something like `buildx prune --free-amount` ?

```
buildx prune --free-amount 5GB
```